File: .github/scripts/detect-changes/lib/dockerfile_parser.py
"""
Dockerfile parsing and dependency extraction.

Parses Dockerfile content to extract:
- Base image references (FROM statements)
- HEALTHCHECK instructions and parameters
- External image dependencies (COPY --from)
- Validation rules (no ARG in FROM)
- Base image upstream information
"""

from typing import Dict, List, Optional
from dockerfile_parse import DockerfileParser


class ValidationError(Exception):
    """Raised when Dockerfile validation fails."""
    pass


def parse_from_lines(dockerfile_content: str) -> List[Dict[str, Optional[str]]]:
    """
    Extract all FROM lines from a Dockerfile.

    Args:
        dockerfile_content: The Dockerfile content as a string

    Returns:
        List of dictionaries with keys:
        - 'image': The base image name (required)
        - 'stage': The stage name if using AS keyword (optional)
        - 'platform': The platform if using --platform flag (optional)

    Example:
        >>> parse_from_lines('FROM node:18 AS base')
        [{'image': 'node:18', 'stage': 'base', 'platform': None}]
    """
    if not dockerfile_content or not dockerfile_content.strip():
        return []

    parser = DockerfileParser()
    parser.content = dockerfile_content

    result = []

    for instruction in parser.structure:
        if instruction['instruction'].upper() == 'FROM':
            value = instruction['value']

            # Skip malformed FROM lines
            if not value or not value.strip():
                continue

            parts = value.split()

            # Handle --platform flag
            platform = None
            image_idx = 0

            if parts and parts[0].startswith('--platform='):
                platform = parts[0].split('=', 1)[1]
                image_idx = 1

            if len(parts) <= image_idx:
                continue  # Malformed

            image = parts[image_idx]

            # Handle AS keyword for stage names
            stage = None
            if len(parts) > image_idx + 1 and parts[image_idx + 1].upper() == 'AS':
                if len(parts) > image_idx + 2:
                    stage = parts[image_idx + 2]

            result.append({
                'image': image,
                'stage': stage,
                'platform': platform
            })

    return result


def extract_final_stage_base(dockerfile_content: str) -> Optional[str]:
    """
    Extract the ultimate external base image for the final stage.

    Follows internal stage references to find the actual external base image.

    Args:
        dockerfile_content: The Dockerfile content as a string

    Returns:
        The external base image used by the final stage, or None if not found

    Example:
        >>> extract_final_stage_base('FROM node:18 AS base\\nFROM base AS final')
        'node:18'
    """
    from_lines = parse_from_lines(dockerfile_content)

    if not from_lines:
        return None

    # Build a map of stage names to their base images
    stage_map = {}
    for from_line in from_lines:
        if from_line['stage']:
            stage_map[from_line['stage']] = from_line['image']

    # Get the final stage's image
    final_image = from_lines[-1]['image']

    # Follow the chain to find the external base
    visited = set()
    current = final_image

    while current in stage_map and current not in visited:
        visited.add(current)
        current = stage_map[current]

    return current


def has_healthcheck(dockerfile_content: str) -> bool:
    """
    Check if the final stage has a HEALTHCHECK instruction.

    Args:
        dockerfile_content: The Dockerfile content as a string

    Returns:
        True if HEALTHCHECK exists in final stage, False otherwise

    Note:
        HEALTHCHECK NONE is treated as disabled (returns False)
    """
    if not dockerfile_content:
        return False

    parser = DockerfileParser()
    parser.content = dockerfile_content

    # Find the last FROM instruction to identify final stage
    last_from_idx = -1
    for i, instruction in enumerate(parser.structure):
        if instruction['instruction'].upper() == 'FROM':
            last_from_idx = i

    if last_from_idx == -1:
        return False

    # Look for HEALTHCHECK after the last FROM
    for i in range(last_from_idx + 1, len(parser.structure)):
        instruction = parser.structure[i]
        if instruction['instruction'].upper() == 'HEALTHCHECK':
            value = instruction['value'].strip()
            # HEALTHCHECK NONE disables health checks
            if value.upper() == 'NONE':
                return False
            return True

    return False


def parse_healthcheck_params(dockerfile_content: str) -> Optional[Dict[str, Optional[str]]]:
    """
    Extract HEALTHCHECK parameters from the final stage.

    Args:
        dockerfile_content: The Dockerfile content as a string

    Returns:
        Dictionary with keys: interval, timeout, start_period, retries, cmd
        Returns None if no HEALTHCHECK instruction found

    Example:
        >>> parse_healthcheck_params('FROM node\\nHEALTHCHECK --interval=30s CMD echo ok')
        {'interval': '30s', 'timeout': None, 'start_period': None, 'retries': None, 'cmd': 'echo ok'}
    """
    if not dockerfile_content:
        return None

    parser = DockerfileParser()
    parser.content = dockerfile_content

    # Find the last FROM instruction to identify final stage
    last_from_idx = -1
    for i, instruction in enumerate(parser.structure):
        if instruction['instruction'].upper() == 'FROM':
            last_from_idx = i

    if last_from_idx == -1:
        return None

    # Look for HEALTHCHECK after the last FROM
    for i in range(last_from_idx + 1, len(parser.structure)):
        instruction = parser.structure[i]
        if instruction['instruction'].upper() == 'HEALTHCHECK':
            value = instruction['value'].strip()

            # HEALTHCHECK NONE has no parameters
            if value.upper() == 'NONE':
                return None

            result = {
                'interval': None,
                'timeout': None,
                'start_period': None,
                'retries': None,
                'cmd': None
            }

            # Parse the value to extract parameters
            parts = []
            current = []
            in_quotes = False

            for char in value:
                if char in ('"', "'"):
                    in_quotes = not in_quotes
                    current.append(char)
                elif char == ' ' and not in_quotes:
                    if current:
                        parts.append(''.join(current))
                        current = []
                else:
                    current.append(char)

            if current:
                parts.append(''.join(current))

            # Extract flags and CMD
            cmd_parts = []
            i = 0
            while i < len(parts):
                part = parts[i]

                if part.startswith('--interval='):
                    result['interval'] = part.split('=', 1)[1]
                elif part.startswith('--timeout='):
                    result['timeout'] = part.split('=', 1)[1]
                elif part.startswith('--start-period='):
                    result['start_period'] = part.split('=', 1)[1]
                elif part.startswith('--retries='):
                    result['retries'] = part.split('=', 1)[1]
                elif part.upper() == 'CMD':
                    # Everything after CMD is the command
                    cmd_parts = parts[i+1:]
                    break
                else:
                    cmd_parts.append(part)

                i += 1

            if cmd_parts:
                result['cmd'] = ' '.join(cmd_parts)

            return result

    return None


def extract_copy_from_external(dockerfile_content: str) -> List[str]:
    """
    Extract external images used in COPY --from statements.

    Internal stage references are filtered out.

    Args:
        dockerfile_content: The Dockerfile content as a string

    Returns:
        List of unique external image names

    Example:
        >>> extract_copy_from_external('COPY --from=node:18 /app /app')
        ['node:18']
    """
    if not dockerfile_content:
        return []

    parser = DockerfileParser()
    parser.content = dockerfile_content

    # Get all stage names to filter them out
    from_lines = parse_from_lines(dockerfile_content)
    stage_names = {line['stage'] for line in from_lines if line['stage']}

    external_images = set()

    for instruction in parser.structure:
        if instruction['instruction'].upper() == 'COPY':
            value = instruction['value']

            # Look for --from= flag
            if '--from=' in value:
                # Extract the image name from --from=image
                parts = value.split()
                for part in parts:
                    if part.startswith('--from='):
                        image = part.split('=', 1)[1]
                        # Only include if it's not an internal stage
                        if image not in stage_names:
                            external_images.add(image)
                        break

    return list(external_images)


def validate_no_arg_in_from(dockerfile_content: str) -> None:
    """
    Validate that FROM lines don't use ARG variables.

    Raises ValidationError if variable substitution is detected in FROM.

    Args:
        dockerfile_content: The Dockerfile content as a string

    Raises:
        ValidationError: If ARG variable is used in FROM line

    Example:
        >>> validate_no_arg_in_from('FROM node:18')  # OK
        >>> validate_no_arg_in_from('FROM node:${VERSION}')  # Raises
    """
    from_lines = parse_from_lines(dockerfile_content)

    for from_line in from_lines:
        image = from_line['image']

        # Check for variable substitution patterns
        if '$' in image:
            raise ValidationError(
                f"FROM line contains variable substitution: {image}. "
                "ARG variables in FROM statements are not allowed."
            )


def parse_base_image_dockerfile(dockerfile_content: str) -> Optional[Dict[str, Optional[str]]]:
    """
    Parse a base image Dockerfile to extract upstream image information.

    This is used for base-images/ directory Dockerfiles to track upstream images.

    Args:
        dockerfile_content: The Dockerfile content as a string

    Returns:
        Dictionary with keys:
        - 'upstream_image': Full upstream image reference
        - 'image_name': Just the image name (without registry/tag)
        - 'version_tag': The version tag, or None/latest if not specified

    Example:
        >>> parse_base_image_dockerfile('FROM node:18.20.8-alpine')
        {'upstream_image': 'node:18.20.8-alpine', 'image_name': 'node', 'version_tag': '18.20.8-alpine'}
    """
    from_lines = parse_from_lines(dockerfile_content)

    if not from_lines:
        return None

    # Use the first external FROM (skip internal stage references)
    upstream_image = None
    for from_line in from_lines:
        image = from_line['image']
        # If it doesn't look like a stage reference (no dots, slashes, or colons typically means it's a stage)
        # But actually, we should just use the first one since base images are simple
        upstream_image = image
        break

    if not upstream_image:
        return None

    # Parse the image name and tag
    # Format: [registry/]image[:tag]

    # Remove registry if present
    image_without_registry = upstream_image
    if '/' in upstream_image:
        parts = upstream_image.split('/')
        # Check if first part looks like a registry (has dot or is localhost)
        if '.' in parts[0] or ':' in parts[0] or parts[0] == 'localhost':
            # This is a registry, remove it
            image_without_registry = '/'.join(parts[1:])
        else:
            # Not a registry, keep as is
            image_without_registry = upstream_image

    # Extract version tag
    if ':' in image_without_registry:
        image_name, version_tag = image_without_registry.rsplit(':', 1)
    else:
        image_name = image_without_registry
        version_tag = None

    # For images with registry, use the part after registry for image_name
    if '/' in upstream_image:
        parts = upstream_image.split('/')
        if '.' in parts[0] or ':' in parts[0] or parts[0] == 'localhost':
            # Has registry
            rest = '/'.join(parts[1:])
            if ':' in rest:
                image_name = rest.rsplit(':', 1)[0]
            else:
                image_name = rest

    return {
        'upstream_image': upstream_image,
        'image_name': image_name,
        'version_tag': version_tag
    }

File: .github/scripts/detect-changes/src/lib/dockerfile-parser.ts
/**
 * Dockerfile parsing and dependency extraction.
 *
 * Parses Dockerfile content to extract:
 * - Base image references (FROM statements)
 * - HEALTHCHECK instructions and parameters
 * - External image dependencies (COPY --from)
 * - Validation rules (no ARG in FROM)
 * - Base image upstream information
 */

import { DockerfileParser, From, Healthcheck, Copy } from 'dockerfile-ast';

/**
 * Custom error class for Dockerfile validation failures.
 */
export class ValidationError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'ValidationError';
  }
}

/**
 * Represents a FROM instruction with parsed components.
 */
export interface FromInstruction {
  /** The base image name */
  image: string;
  /** The stage name if using AS keyword (optional) */
  stage: string | null;
  /** The platform if using --platform flag (optional) */
  platform: string | null;
}

/**
 * Represents HEALTHCHECK parameters extracted from Dockerfile.
 */
export interface HealthcheckParams {
  /** Health check interval duration */
  interval: string | null;
  /** Health check timeout duration */
  timeout: string | null;
  /** Start period before health checks begin */
  start_period: string | null;
  /** Number of retries before marking unhealthy */
  retries: string | null;
  /** The command to execute for health check */
  cmd: string | null;
}

/**
 * Represents parsed base image information.
 */
export interface BaseImageInfo {
  /** Full upstream image reference */
  upstream_image: string;
  /** Just the image name (without registry/tag) */
  image_name: string;
  /** The version tag, or null if not specified */
  version_tag: string | null;
}

/**
 * Extract all FROM lines from a Dockerfile.
 *
 * @param dockerfileContent - The Dockerfile content as a string
 * @returns List of FROM instructions with image, stage, and platform information
 *
 * @example
 * ```typescript
 * parseFromLines('FROM node:18 AS base')
 * // Returns: [{ image: 'node:18', stage: 'base', platform: null }]
 * ```
 */
export function parseFromLines(dockerfileContent: string): FromInstruction[] {
  if (!dockerfileContent || !dockerfileContent.trim()) {
    return [];
  }

  const parser = DockerfileParser.parse(dockerfileContent);
  const result: FromInstruction[] = [];

  for (const instruction of parser.getInstructions()) {
    if (instruction.getKeyword().toUpperCase() === 'FROM') {
      const fromInstruction = instruction as From;

      // Get the full image string including registry
      const imageRange = fromInstruction.getImageRange();
      if (!imageRange) {
        continue; // Malformed
      }

      // Extract the full image name from the content
      const content = dockerfileContent.split('\n');
      const startLine = imageRange.start.line;
      const startChar = imageRange.start.character;
      const endLine = imageRange.end.line;
      const endChar = imageRange.end.character;

      let image: string;
      if (startLine === endLine) {
        image = content[startLine].substring(startChar, endChar);
      } else {
        // Multi-line (unlikely for image name, but handle it)
        const lines = content.slice(startLine, endLine + 1);
        lines[0] = lines[0].substring(startChar);
        lines[lines.length - 1] = lines[lines.length - 1].substring(0, endChar);
        image = lines.join('\n');
      }

      // Get build stage name
      const stage = fromInstruction.getBuildStage() || null;

      // Get platform flag
      const platformFlag = fromInstruction.getPlatformFlag();
      const platform = platformFlag ? platformFlag.getValue() : null;

      result.push({
        image,
        stage,
        platform,
      });
    }
  }

  return result;
}

/**
 * Extract the ultimate external base image for the final stage.
 *
 * Follows internal stage references to find the actual external base image.
 *
 * @param dockerfileContent - The Dockerfile content as a string
 * @returns The external base image used by the final stage, or null if not found
 *
 * @example
 * ```typescript
 * extractFinalStageBase('FROM node:18 AS base\nFROM base AS final')
 * // Returns: 'node:18'
 * ```
 */
export function extractFinalStageBase(dockerfileContent: string): string | null {
  const fromLines = parseFromLines(dockerfileContent);

  if (fromLines.length === 0) {
    return null;
  }

  // Build a map of stage names to their base images
  const stageMap = new Map<string, string>();
  for (const fromLine of fromLines) {
    if (fromLine.stage) {
      stageMap.set(fromLine.stage, fromLine.image);
    }
  }

  // Get the final stage's image
  const finalImage = fromLines[fromLines.length - 1].image;

  // Follow the chain to find the external base
  const visited = new Set<string>();
  let current = finalImage;

  while (stageMap.has(current) && !visited.has(current)) {
    visited.add(current);
    current = stageMap.get(current)!;
  }

  return current;
}

/**
 * Check if the final stage has a HEALTHCHECK instruction.
 *
 * @param dockerfileContent - The Dockerfile content as a string
 * @returns True if HEALTHCHECK exists in final stage, False otherwise
 *
 * @remarks
 * HEALTHCHECK NONE is treated as disabled (returns false)
 */
export function hasHealthcheck(dockerfileContent: string): boolean {
  if (!dockerfileContent) {
    return false;
  }

  const parser = DockerfileParser.parse(dockerfileContent);
  const instructions = parser.getInstructions();

  // Find the last FROM instruction to identify final stage
  let lastFromIdx = -1;
  for (let i = 0; i < instructions.length; i++) {
    if (instructions[i].getKeyword().toUpperCase() === 'FROM') {
      lastFromIdx = i;
    }
  }

  if (lastFromIdx === -1) {
    return false;
  }

  // Look for HEALTHCHECK after the last FROM
  for (let i = lastFromIdx + 1; i < instructions.length; i++) {
    const instruction = instructions[i];
    if (instruction.getKeyword().toUpperCase() === 'HEALTHCHECK') {
      const healthcheck = instruction as Healthcheck;
      const value = healthcheck.getArgumentsContent()?.trim() || '';
      // HEALTHCHECK NONE disables health checks
      if (value.toUpperCase() === 'NONE') {
        return false;
      }
      return true;
    }
  }

  return false;
}

/**
 * Extract HEALTHCHECK parameters from the final stage.
 *
 * @param dockerfileContent - The Dockerfile content as a string
 * @returns Dictionary with interval, timeout, start_period, retries, and cmd, or null if not found
 *
 * @example
 * ```typescript
 * parseHealthcheckParams('FROM node\nHEALTHCHECK --interval=30s CMD echo ok')
 * // Returns: { interval: '30s', timeout: null, start_period: null, retries: null, cmd: 'echo ok' }
 * ```
 */
export function parseHealthcheckParams(dockerfileContent: string): HealthcheckParams | null {
  if (!dockerfileContent) {
    return null;
  }

  const parser = DockerfileParser.parse(dockerfileContent);
  const instructions = parser.getInstructions();

  // Find the last FROM instruction to identify final stage
  let lastFromIdx = -1;
  for (let i = 0; i < instructions.length; i++) {
    if (instructions[i].getKeyword().toUpperCase() === 'FROM') {
      lastFromIdx = i;
    }
  }

  if (lastFromIdx === -1) {
    return null;
  }

  // Look for HEALTHCHECK after the last FROM
  for (let i = lastFromIdx + 1; i < instructions.length; i++) {
    const instruction = instructions[i];
    if (instruction.getKeyword().toUpperCase() === 'HEALTHCHECK') {
      const healthcheck = instruction as Healthcheck;

      // Get the subcommand (CMD or NONE)
      const subcommand = healthcheck.getSubcommand();
      if (subcommand && subcommand.getValue().toUpperCase() === 'NONE') {
        return null;
      }

      const result: HealthcheckParams = {
        interval: null,
        timeout: null,
        start_period: null,
        retries: null,
        cmd: null,
      };

      // Extract flags using the getFlags() method
      const flags = healthcheck.getFlags();
      if (flags) {
        for (const flag of flags) {
          const name = flag.getName();
          const value = flag.getValue();

          if (name === 'interval') {
            result.interval = value;
          } else if (name === 'timeout') {
            result.timeout = value;
          } else if (name === 'start-period') {
            result.start_period = value;
          } else if (name === 'retries') {
            result.retries = value;
          }
        }
      }

      // Get the CMD part
      const argsContent = healthcheck.getArgumentsContent()?.trim();
      if (argsContent && argsContent.toUpperCase().startsWith('CMD ')) {
        result.cmd = argsContent.substring(4).trim();
      } else if (argsContent && !argsContent.toUpperCase().startsWith('NONE')) {
        // If there's content but doesn't start with CMD, it might still be the command
        result.cmd = argsContent;
      }

      return result;
    }
  }

  return null;
}

/**
 * Extract external images used in COPY --from statements.
 *
 * Internal stage references are filtered out.
 *
 * @param dockerfileContent - The Dockerfile content as a string
 * @returns List of unique external image names
 *
 * @example
 * ```typescript
 * extractCopyFromExternal('COPY --from=node:18 /app /app')
 * // Returns: ['node:18']
 * ```
 */
export function extractCopyFromExternal(dockerfileContent: string): string[] {
  if (!dockerfileContent) {
    return [];
  }

  const parser = DockerfileParser.parse(dockerfileContent);

  // Get all stage names to filter them out
  const fromLines = parseFromLines(dockerfileContent);
  const stageNames = new Set(fromLines.filter((line) => line.stage).map((line) => line.stage!));

  const externalImages = new Set<string>();

  for (const instruction of parser.getInstructions()) {
    if (instruction.getKeyword().toUpperCase() === 'COPY') {
      const copyInstruction = instruction as Copy;

      // Use the getFromFlag() method to extract --from value
      const fromFlag = copyInstruction.getFromFlag();
      if (fromFlag) {
        const image = fromFlag.getValue();
        // Only include if it's not an internal stage
        if (image && !stageNames.has(image)) {
          externalImages.add(image);
        }
      }
    }
  }

  return Array.from(externalImages);
}

/**
 * Validate that FROM lines don't use ARG variables.
 *
 * Throws ValidationError if variable substitution is detected in FROM.
 *
 * @param dockerfileContent - The Dockerfile content as a string
 * @throws {ValidationError} If ARG variable is used in FROM line
 *
 * @example
 * ```typescript
 * validateNoArgInFrom('FROM node:18')  // OK
 * validateNoArgInFrom('FROM node:${VERSION}')  // Throws ValidationError
 * ```
 */
export function validateNoArgInFrom(dockerfileContent: string): void {
  const fromLines = parseFromLines(dockerfileContent);

  for (const fromLine of fromLines) {
    const image = fromLine.image;

    // Check for variable substitution patterns
    if (image.includes('$')) {
      throw new ValidationError(
        `FROM line contains variable substitution: ${image}. ARG variables in FROM statements are not allowed.`
      );
    }
  }
}

/**
 * Parse a base image Dockerfile to extract upstream image information.
 *
 * This is used for base-images/ directory Dockerfiles to track upstream images.
 *
 * @param dockerfileContent - The Dockerfile content as a string
 * @returns Dictionary with upstream_image, image_name, and version_tag, or null if not found
 *
 * @example
 * ```typescript
 * parseBaseDockerfile('FROM node:18.20.8-alpine')
 * // Returns: { upstream_image: 'node:18.20.8-alpine', image_name: 'node', version_tag: '18.20.8-alpine' }
 * ```
 */
export function parseBaseDockerfile(dockerfileContent: string): BaseImageInfo | null {
  const fromLines = parseFromLines(dockerfileContent);

  if (fromLines.length === 0) {
    return null;
  }

  // Use the first external FROM (skip internal stage references)
  let upstreamImage: string | null = null;
  for (const fromLine of fromLines) {
    const image = fromLine.image;
    // If it doesn't look like a stage reference (no dots, slashes, or colons typically means it's a stage)
    // But actually, we should just use the first one since base images are simple
    upstreamImage = image;
    break;
  }

  if (!upstreamImage) {
    return null;
  }

  // Parse the image name and tag
  // Format: [registry/]image[:tag]

  // Remove registry if present
  let imageWithoutRegistry = upstreamImage;
  if (upstreamImage.includes('/')) {
    const parts = upstreamImage.split('/');
    // Check if first part looks like a registry (has dot or is localhost)
    if (parts[0].includes('.') || parts[0].includes(':') || parts[0] === 'localhost') {
      // This is a registry, remove it
      imageWithoutRegistry = parts.slice(1).join('/');
    } else {
      // Not a registry, keep as is
      imageWithoutRegistry = upstreamImage;
    }
  }

  // Extract version tag
  let imageName: string;
  let versionTag: string | null;

  if (imageWithoutRegistry.includes(':')) {
    const lastColonIdx = imageWithoutRegistry.lastIndexOf(':');
    imageName = imageWithoutRegistry.substring(0, lastColonIdx);
    versionTag = imageWithoutRegistry.substring(lastColonIdx + 1);
  } else {
    imageName = imageWithoutRegistry;
    versionTag = null;
  }

  // For images with registry, use the part after registry for image_name
  if (upstreamImage.includes('/')) {
    const parts = upstreamImage.split('/');
    if (parts[0].includes('.') || parts[0].includes(':') || parts[0] === 'localhost') {
      // Has registry
      const rest = parts.slice(1).join('/');
      if (rest.includes(':')) {
        const lastColonIdx = rest.lastIndexOf(':');
        imageName = rest.substring(0, lastColonIdx);
      } else {
        imageName = rest;
      }
    }
  }

  return {
    upstream_image: upstreamImage,
    image_name: imageName,
    version_tag: versionTag,
  };
}
